src/persist/src/file.rs

```diff
@@ -182,6 +183,179 @@ impl Persister for FilePersister {
+    /// include anything that would help debug an unexpected LOCK file, such as
+    /// version, ip, worker number, etc.
+    pub fn new<P: AsRef<Path>>(base_dir: P, lock_info: &[u8]) -> Result<Self, Error> {
+        let base_dir = base_dir.as_ref();
+        fs::create_dir_all(&base_dir)?;
+        {
+            let lockfile_path = Self::lockfile_path(&base_dir);
+            let mut lockfile = OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(lockfile_path)?;
+            lockfile.write_all(&lock_info)?;
+            lockfile.sync_all()?;
+        }
+        let buffer_path = Self::buffer_path(&base_dir);
+        let buffer_file = OpenOptions::new()
+            .append(true)
+            .read(true)
+            .create(true)
+            .open(&buffer_path)?;
+
+        let metadata_path = Self::metadata_path(&base_dir);
+        let mut metadata_file = OpenOptions::new()
+            .write(true)
+            .read(true)
+            .create(true)
+            .open(&metadata_path)?;
+
+        // Retrieve the sequence number metadata from the file
+        let mut bytes = Vec::new();
+        metadata_file.read_to_end(&mut bytes)?;
+        let seqno = if bytes.len() != 0 {
+            assert!(bytes.len() == 16);
+            let mut bytes = &bytes[..];
+            let mut buf = [0u8; 8];
+            // Decode sequence number
+            bytes.read_exact(&mut buf)?;
+            let seqno_start = SeqNo(u64::from_le_bytes(buf));
+            bytes.read_exact(&mut buf)?;
+            let seqno_end = SeqNo(u64::from_le_bytes(buf));
+            seqno_start..seqno_end
+        } else {
+            let bytes = [0u8; 16];
+            metadata_file.write_all(&bytes)?;
+            SeqNo(0)..SeqNo(0)
+        };
+
+        Ok(FileBuffer {
+            dataz: Arc::new(Mutex::new((buffer_file, metadata_file))),
+            seqno,
+            buf: Vec::new(),
+        })
+    }
+
+    fn lockfile_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::LOCKFILE_PATH)
+    }
+
+    fn buffer_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::BUFFER_PATH)
+    }
+
+    fn metadata_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::METADATA_PATH)
+    }
+}
+
+impl Buffer for FileBuffer {
+    fn write_sync(&mut self, buf: Vec<u8>) -> Result<SeqNo, Error> {
+        self.buf.clear();
+        let write_seqno = self.seqno.end;
+        let len = buf.len() as u64;
+
+        self.buf.extend(&write_seqno.0.to_le_bytes());
+        self.buf.extend(&len.to_le_bytes());
+        self.buf.extend(buf);
+        self.buf.shrink_to_fit();
+
+        let mut guard = self.dataz.lock()?;
+        guard.0.write_all(&self.buf)?;
+        guard.0.sync_all()?;
+        self.seqno = self.seqno.start..SeqNo(self.seqno.end.0 + 1);
+        guard.1.seek(SeekFrom::Start(8))?;
+        guard.1.write_all(&self.seqno.end.0.to_le_bytes())?;
```

Comment:
need to sync here as well

Reply:
also the data and metadata writes are not atomic -- not sure how to resolve that

Reply:
after a walk its clear -- we don't actually need to write down the seqno upper bound -- we happen to do so in order to then read it back in on startup, but we could instead just keep the lower bound. At that point, we only need to write to metadata on truncate. The tradeoff then is -- on restart we need to scan the whole buffer to see what the seqno upper bound is but instead of being dumb we could be just a little bit clever and reorder the binary layout to be

binary data length
binary data
seqno

at which point we could grab the last upper bound with a seek from the end.